L1Trigger/Phase2L1ParticleFlow/src/BitwisePFAlgo.cc

```diff
@@ -0,0 +1,204 @@
+  if (algo == "pfalgo3") {
+    algo_ = algo3;
+    config_ = new pfalgo3_config(bitwiseConfig.getParameter<uint32_t>("NTRACK"),
+                                 bitwiseConfig.getParameter<uint32_t>("NEMCALO"),
+                                 bitwiseConfig.getParameter<uint32_t>("NCALO"),
+                                 bitwiseConfig.getParameter<uint32_t>("NMU"),
+                                 bitwiseConfig.getParameter<uint32_t>("NPHOTON"),
+                                 bitwiseConfig.getParameter<uint32_t>("NSELCALO"),
+                                 bitwiseConfig.getParameter<uint32_t>("NALLNEUTRAL"),
+                                 bitwiseConfig.getParameter<uint32_t>("DR2MAX_TK_MU"),
+                                 bitwiseConfig.getParameter<uint32_t>("DR2MAX_TK_EM"),
+                                 bitwiseConfig.getParameter<uint32_t>("DR2MAX_EM_CALO"),
+                                 bitwiseConfig.getParameter<uint32_t>("DR2MAX_TK_CALO"),
+                                 bitwiseConfig.getParameter<uint32_t>("TK_MAXINVPT_LOOSE"),
+                                 bitwiseConfig.getParameter<uint32_t>("TK_MAXINVPT_TIGHT"));
+    pfalgo3_ref_set_debug(debug_);
+  } else if (algo == "pfalgo2hgc") {
+    algo_ = algo2hgc;
+    config_ = new pfalgo_config(bitwiseConfig.getParameter<uint32_t>("NTRACK"),
+                                bitwiseConfig.getParameter<uint32_t>("NCALO"),
+                                bitwiseConfig.getParameter<uint32_t>("NMU"),
+                                bitwiseConfig.getParameter<uint32_t>("NSELCALO"),
+                                bitwiseConfig.getParameter<uint32_t>("DR2MAX_TK_MU"),
+                                bitwiseConfig.getParameter<uint32_t>("DR2MAX_TK_CALO"),
+                                bitwiseConfig.getParameter<uint32_t>("TK_MAXINVPT_LOOSE"),
+                                bitwiseConfig.getParameter<uint32_t>("TK_MAXINVPT_TIGHT"));
+    pfalgo2hgc_ref_set_debug(debug_);
+  } else {
+    throw cms::Exception("Configuration", "Unsupported bitwiseAlgo " + algo);
+  }
+}
+
+BitwisePFAlgo::~BitwisePFAlgo() {
+  switch (algo_) {
+    case algo3:
+      delete (static_cast<pfalgo3_config *>(config_));
+      break;
+    default:
+      delete config_;
+  }
+  config_ = nullptr;
+}
+
+void BitwisePFAlgo::runPF(Region &r) const {
+  initRegion(r);
+
+  std::unique_ptr<HadCaloObj[]> calo(new HadCaloObj[config_->nCALO]);
+  std::unique_ptr<TkObj[]> track(new TkObj[config_->nTRACK]);
+  std::unique_ptr<MuObj[]> mu(new MuObj[config_->nMU]);
+  std::unique_ptr<PFChargedObj[]> outch(new PFChargedObj[config_->nTRACK]);
+  std::unique_ptr<PFNeutralObj[]> outne(new PFNeutralObj[config_->nSELCALO]);
+  std::unique_ptr<PFChargedObj[]> outmu(new PFChargedObj[config_->nMU]);
+
+  dpf2fw::convert(config_->nTRACK, r.track, track.get());
+  dpf2fw::convert(config_->nCALO, r.calo, calo.get());
+  dpf2fw::convert(config_->nMU, r.muon, mu.get());
+
+  if (debug_) {
+    printf(
```

Comment:
printf is not allowed, use MessageLogger i.e. LogDebug

Reply:
It's inside a debug block, used only for debugging, and producing properly aligned output with c++ streaming requires much more code and is less readable.

Reply:
it's not thread safe, therefore not allowed in code that runs in production.

Reply:
if you really want the printf formatting, use sprintf and pass the resulting string to the message logger.